### PR TITLE
Ensure Endfilter basic block is preserved even if technically unreachable.

### LINKED ISSUE #67494: Invalid IL produced for an always-throwing filter expression
**Version Used**: Sharplab's `main 25 Jan 2023`.

**Steps to Reproduce**:

Sharplab link: https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmOATEBqAPgAQBYACABQCcB7YAGwFMwAKASgFgAoAbw+N+IXIBPHn27s+xAL4jeAYwgJZAC2IB3JXRjFGAgK51iAfn5Kqq4jF00axEP3L6240TKmuJHSUA

Jit-compiling `Problem` below (in Debug) will crash the runtime with an `InvalidProgramException`:
```cs
void Problem()
{
    try
    {
    }
    catch when (true ? throw null : true)
    {
    }
}
```
The IL produced for the filter does not conclude with `endfilter`, even as it must:
```
ECMA 335

I.12.4.2.8.2.5 endfilter: 

1. Shall appear as the lexically last instruction in the filter. 
[Note: The endfilter is required even if no control-flow path reaches it. This can happen if, for 
example, the filter does a throw. end note] 
```

**Expected Behavior**: correct IL is produced.

**Actual Behavior**: incorrect IL is produced.

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #67494.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs [hunk 1]
```diff
@@ -3758,5 +3759,492 @@ .maxstack  1
...
+
+    static System.Exception M2()
+    {
+        System.Console.Write("M2");
+        return new System.NotSupportedException();
+    }
+
+    static void M3()
+    {
+        System.Console.Write("M2");
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs:3816]: nit: The method name in this call doesn't match the containing method. Consider removing the call if it is not used by the test.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs (hunk 1 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs:3816]:
```suggestion
        System.Console.Write("M3");
```

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs [hunk 2]
```diff
@@ -3758,5 +3759,492 @@ .maxstack  1
...
+    {
+        System.Console.Write("M4");
+    }
+}
+""";
+
+            var verifier = CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: "M1M4M1M2Catch",
+                // False PEVerify failure:
+                // PEVerify failed for assembly
+                // [ : C::M][mdToken=0x6000004][offset 0x0000000E] Stack not empty when leaving an exception filter.
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs:3829]: I think the PEVerify failures can be verified programmatically by something like `Verification.FailsPEVerify with { PEVerifyMessage = "..." }`.

**Review by jjonescz — APPROVED** [review]

**Pull request merged**